### PR TITLE
[#162393] Avoid creation of service and item price policies with usage_rate

### PR DESCRIPTION
**Pull request by diebas**
# Release Notes

There's a service that automatically creates price_policies with a usage_rate value of 0. This is not totally real as some price_policies aren't meant to have usage_rate at all which may result in future bugs.

Change the code to reflect reality, add tests, and refactor a little to reuse some code.

### CODE REVIEW

File: spec/support/shared_examples/create_service_with_price_rule_shared_examples.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+RSpec.shared_examples "creates a product with billing mode" do |product_type, billing_mode|
```
Comment (LeticiaErrandonea) [inline spec/support/shared_examples/create_service_with_price_rule_shared_examples.rb:1]: nit: please add `# frozen_string_literal: true`

nit 2: please add newline at the end of the file

nit 3: the filename should be something like "create_product_with_price_rule_shared_examples.rb", right? We're not only creating services

File: spec/support/shared_examples/create_service_with_price_rule_shared_examples.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  let(:logged_in_user) { administrator }
+
+  it "can create a #{product_type}, which automatically creates a price rule" do
+    visit facility_products_path(facility)
+    all("a", text: product_type.capitalize)[0].click
```
Comment (LeticiaErrandonea) [inline spec/support/shared_examples/create_service_with_price_rule_shared_examples.rb:6]: Why do we need to get `all` and then select the first one?

File: spec/support/shared_examples/create_service_with_price_rule_shared_examples.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    click_link "Add #{product_type.capitalize}"
+
+    fill_in "Name", with: "My new Product", match: :first
+    fill_in "URL Name", with: "new-service"
+    select "Required", from: "#{product_type}[user_notes_field_mode]"
+    select billing_mode, from: "#{product_type}[billing_mode]"
```
Comment (LeticiaErrandonea) [inline spec/support/shared_examples/create_service_with_price_rule_shared_examples.rb:12]: Can we do `choose billing_mode`?

File: spec/support/shared_examples/create_service_with_price_rule_shared_examples.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+    click_button "Create"
+    click_on "Pricing"
```
Comment (LeticiaErrandonea) [inline spec/support/shared_examples/create_service_with_price_rule_shared_examples.rb:15]: Between these two I'd check that the product was created

File: spec/support/shared_examples/create_service_with_price_rule_shared_examples.rb [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+    expect(page).to have_content "#{PriceGroup.nonbillable} (#{PriceGroup.nonbillable.type_string}) $0"
```
Comment (LeticiaErrandonea) [inline spec/support/shared_examples/create_service_with_price_rule_shared_examples.rb:17]: If we're only creating nonbillable, do we need the `billing_mode` passed in?

**Review by LeticiaErrandonea — COMMENTED** [review]
Looking good!

Left some comments with questions

File: app/services/price_policy_builder.rb [hunk 6]
```diff
@@ -154,4 +124,25 @@ def price_policies_for_start_date
...
+  def self.create_price_policy_for(product, price_group)
+    usage_rate = ["Service", "Item"].include?(product.type) ? nil : 0
+
+    PricePolicy.create(
+      type: "#{product.type}PricePolicy",
+      product: product,
+      start_date: 1.month.ago,
+      expire_date: 75.years.from_now,
+      price_group: price_group,
+      usage_rate: usage_rate,
```
Comment (joaquinco) [inline app/services/price_policy_builder.rb:136]: nit: we can use the hash shorthand syntax here

File: spec/support/shared_examples/create_service_with_price_rule_shared_examples.rb (hunk 2 above)
Comment (diebas) [reply spec/support/shared_examples/create_service_with_price_rule_shared_examples.rb:6]: The reason behind this is that when the product type is service, there's another link in the menu called "Timed Services" that interferes when calling `click_link "Services"`. With this code, I find all the links that contain the word "Services" (there are only 2) and select the first one, which is the link I need to visit.

File: spec/support/shared_examples/create_service_with_price_rule_shared_examples.rb (hunk 3 above)
Comment (diebas) [reply spec/support/shared_examples/create_service_with_price_rule_shared_examples.rb:12]: I don't think so, as this is a drop-down, not a radio button

File: app/services/price_policy_builder.rb [hunk 7]
```diff
@@ -154,4 +124,25 @@ def price_policies_for_start_date
     product.price_policies.for_date(start_date)
   end
 
+  def self.create_price_policy_for(product, price_group)
+    usage_rate = ["Service", "Item"].include?(product.type) ? nil : 0
```
Comment (giladshanan) [inline app/services/price_policy_builder.rb:128]: I'd vote for making this a separate method and adding a simple test for it

File: spec/support/shared_examples/create_service_with_price_rule_shared_examples.rb (hunk 5 above)
Comment (diebas) [reply spec/support/shared_examples/create_service_with_price_rule_shared_examples.rb:17]: We are not only just creating products with nonbillable billing_mode, we are creating skip review too.
As I understand, creating products of both billing modes generates "nonbillable" price groups (base price group with $0 cost). This code checks that the price group is being displayed correctly in the product's price tab.

`PriceGroup.nonbillable ` returns the base price group

```
  def self.nonbillable
    base
  end

  def self.base
    globals.find_by(name: Settings.price_group.name.base)
  end
```

File: spec/support/shared_examples/create_service_with_price_rule_shared_examples.rb (hunk 2 above)
Comment (LeticiaErrandonea) [reply spec/support/shared_examples/create_service_with_price_rule_shared_examples.rb:6]: Oh, I think that you can use something like `click_link("Services", href: facility_services_path(facility))`

File: spec/support/shared_examples/create_service_with_price_rule_shared_examples.rb (hunk 3 above)
Comment (LeticiaErrandonea) [reply spec/support/shared_examples/create_service_with_price_rule_shared_examples.rb:12]: You're right 😅 

What about something like `select billing_mode, from: "Billing Mode"`?

File: spec/services/price_policy_builder_spec.rb [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+require "rails_helper"
+
+RSpec.describe PricePolicyBuilder do
+  let(:price_group) { create(:price_group) }
+  let!(:facility) { create(:setup_facility) }
+
+  shared_examples "a correctly created PricePolicy" do
+    before do
+      allow_any_instance_of(PricePolicy).to receive(:validate).and_return(true)
```
Comment (LeticiaErrandonea) [inline spec/services/price_policy_builder_spec.rb:11]: Why do we need to skip validations?

**Review by LeticiaErrandonea — COMMENTED** [review]
Left a question but other than that I think this is fine 😄

**Review by LeticiaErrandonea — APPROVED** [review]
LGTM! Thanks for the changes!

File: app/services/price_policy_builder.rb [hunk 9]
```diff
@@ -154,4 +124,27 @@ def price_policies_for_start_date
...
+      unit_cost: 0,
+      unit_subsidy: 0,
+      can_purchase: true,
+      charge_for: "reservation",
+      note: "Price rule automatically created because of billing mode"
+    )
+  end
+
+  def self.usage_rate_for(product)
+    usage_rate = ["Service", "Item"].include?(product.type) ? nil : 0
```
Comment (giladshanan) [inline app/services/price_policy_builder.rb:147]:
```suggestion
    ["Service", "Item"].include?(product.type) ? nil : 0
```

File: spec/support/shared_examples/create_product_with_price_rule_shared_examples.rb [hunk 10]
```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+RSpec.shared_examples "creates a product with billing mode" do |product_type, billing_mode|
```
Comment (joaquinco) [inline spec/support/shared_examples/create_product_with_price_rule_shared_examples.rb:3]: I think it if it's used by a single example group we can define these shared examples with the same file

File: app/services/price_policy_builder.rb [hunk 11]
```diff
@@ -154,4 +124,27 @@ def price_policies_for_start_date
     product.price_policies.for_date(start_date)
   end
 
+  def self.create_price_policy_for(product, price_group)
+    PricePolicy.create(
+      type: "#{product.type}PricePolicy",
+      product: product,
```
Comment (joaquinco) [inline app/services/price_policy_builder.rb:130]: nit: hash value can be omitted

**Review by joaquinco — APPROVED** [review]

File: spec/support/shared_examples/create_product_with_price_rule_shared_examples.rb (hunk 10 above)
Comment (diebas) [reply spec/support/shared_examples/create_product_with_price_rule_shared_examples.rb:3]: It's used in 2 different files:

spec/system/admin/creating_a_service_spec.rb
spec/system/admin/creating_an_item_spec.rb

Comment (joaquinco) [reply spec/support/shared_examples/create_product_with_price_rule_shared_examples.rb:3]: I hadn't noticed it

**Pull request merged**